### PR TITLE
Data extraction for rarepokemon - Pokesniper

### PR DESCRIPTION
**Pull request by vivek-sethia**
Data Extraction implemented for getting rare Pokemon appearance details.

First phase listening through api and storing the data into temporary json files( npm run listen --collection=rarePokemon to execute)

Second phase storing the data into the database from temporary files and deleting them. ( npm run update --collection=rarePokemon to execute )

### CODE REVIEW

File: app/models/rarePokemon.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+var mongoose = require('mongoose');
+var Schema   = mongoose.Schema;
+
+var rarePokemonSchema = new Schema({
+    id           : {type: Number},
+    name         : {type: String},
+    latitude   : {type: Number},
+    longitude  : {type: Number},
+    until        : {type: String},
```
Comment (sacdallago) [inline app/models/rarePokemon.js]: `until` as in "seen until"? Why not `Date` instead of `String`?

Comment (vivek-sethia) [inline app/models/rarePokemon.js]: Will do the changes. Thanks for pointing it out.

**Comment by vivek-sethia** [comment]
changes done with respect to 'until' feild.

**Comment by sacdallago** [comment]
@vivek-sethia thanks for the changes!

There is something that still bugs me though: why is there a `seen_until` but no `timestamp` attribute? Basically you are setting an upper bound, but no lower bound.. Maybe what I don't understand is what the data item stands for. Is it an event like "seen here"?

Another note for your group: To represent latitude/longitude you can use your own definitions (as you did), totally fine. **BUT** if you want to stick to some standards, there is http://geojson.org .
In this case you would need to restructure a little bit your documents putting the details in the `properties` key of the GeoJSON object, saving the coordinates as a point (there are Points, LineStrings and Polygons in the specification http://geojson.org/geojson-spec.html#geometry-objects ).

It's just a suggestion! Standardizing data is though always good practice :)

**Comment by vivek-sethia** [comment]
Sample data which is stored in Db : 
{
    "_id": {
        "$oid": "57bac4d2d179f9661c4b23e6"
    },
    "seen_until": {
        "$date": "2016-08-22T09:24:39.000Z"
    },
    "longitude": 9.175103,
    "latitude": 45.4578,
    "name": "Grimer",
    "id": 171723,
    "__v": 0
}

So seen until contains both and time ( until when the pokemon will be available)

**Comment by sacdallago** [comment]
@vivek-sethia mh still not clear to me.

In your example `seen_until` contains only one date. I don't know if this date is in the past, in the future, or what it sands for. Do you mean that this is a data item representing one instance of a sighting of pokemon?

If so, then please remove `until` because it's confusing. `until` in common understanding implies that there is a timeframe, but you have no lower-bound in the data-item itself, so you cannot define any precise time-frame by looking at ONE data-item, just reconstruct it from A SET of data-items :)

Edit:
use the power of markdown! :smiley: Enclose code in triple backthicks (``` x3)  followed by `javascript` to highlight in javascript syntax
https://en.support.wordpress.com/markdown-quick-reference/

**Comment by vivek-sethia** [comment]
until represented from the time when the data was extracted and since this data is going to be later in time, it makes more sense to keep the field as seen or last seen.
Thanks for the correction :)

**Comment by jonas-he** [comment]
@sacdallago Yes, `seen_until` is "representing one instance of a sighting of pokemon" and means that this specific pokemon expires at this date. In general pokemon stay on the map for 15 minutes, which means that we can calculate the `spawn_time` as `seen_until - 15 min`. So there you have your time-frame :smiley:

**Comment by sacdallago** [comment]
Aahwwww now it's clear to me @jonas-he , thanks.

Then, follow up question: do we have the absolute latest point of time for the seeing? Because there is an important difference between `existence in [x-15, x]` and `existence in [x-15 +e, x+e]` where `e in [0, 15]`

**Comment by jonas-he** [comment]
I did some research and it turns out that the `see_until` value of pokesnipers is equal to your `x` as in

> existence in [x-15, x]

Did the verification with https://github.com/mchristopher/PokemonGo-DesktopMap
This is a scanner that uses a reverse engineered version of the API between Pokemon GO app and the backend servers of Niantic. So the data this map represents is the actual data from the game, no prediction, no user submitted sighting ...
Also had the thought to use this as a data source but it would take a lot of resources to get this to a global level since scanning is slow and you need an actual google account or PTC account to run one scanner (and they are at the risk of getting banned) so you would need to generate many accounts and so on....

**Comment by sacdallago** [comment]
wow, cool @jonas-he 
Ok, then it is all fine :)

Edit: and yeah, it's bad that we can't use that very nice but very "illegal" data source

**Comment by vivek-sethia** [comment]
Now spawn time is also stored along with seen_until. Exiting of process after all the data is stored in DB is also handled now. @jonas-he

File: app/controllers/filler/pokemon.js [hunk 2]
```diff
@@ -7,29 +7,21 @@ const PokemonModel = require(__appbase + 'models/pokemon'),
...
-        logger.info('MongoDb Insertion...');
 
         var addPokemon = function (pokemon, callback) {
             PokemonStore.add(pokemon, function (success, data) {
-                if (success != 1) {
-                    logger.success('Success: ' + data.pokemonId);
-                } else {
-                    logger.error('Error:' + data);
-                }
+                logger.success((success != 1) ? 'Error:' + data : 'Success: ' + data.pokemonId);
```
Comment (swathi-ssunder) [inline app/controllers/filler/pokemon.js]: This will cause success message to be logged in either case... 
It is better to revert this change.

File: config.js [hunk 3]
```diff
@@ -19,6 +19,9 @@ module.exports = {
         "token_secret": process.env.ACCESS_TOKEN_SECRET
     },
     "twitterKeyWords" : 'pokemon catch,pokemon saw,pokemon attack,pokemon find,pokemon caught,pokemon attacked,pokemon found,pokemon appeared,#foundPokemon,#caughtPokemon,#pokemongo,a wild appeared until #pokemongo',
+    "pokesniperUrl" : "https://pokesnipers.com/api/v1/pokemon.json?referrer=home",
+    "pokesniperListeningInterval" : 10000, // 10 minutes in milliseconds,
+    "pokemonSpawnTime" : 900 ,// 15 minutes in seconds
```
Comment (swathi-ssunder) [inline config.js:26]: It would be more readable if these details(which are specific to pokesniper) are nested under a 'pokesniper' key.

File: scripts/listen.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+"use strict";
+
+require(__dirname + '/../' + 'constants');
+
+const requested = process.env.npm_config_collection;
```
Comment (swathi-ssunder) [inline scripts/listen.js]: It might be better to initialize this in constants.js and also provide a default value if  'collection' is not passed in the npm configuration.

Comment (swathi-ssunder) [inline scripts/listen.js]: Also, 'requested' seems ambiguous for a variable name. It could be renamed to something more meaningful.

**Comment by vivek-sethia** [comment]
@swathi-ssunder Suggested changes are done.

**Comment by swathi-ssunder** [comment]
The code seems okay to me. I have not pulled and tested though.

**Comment by jonas-he** [comment]
Tested and everything works :+1:
